tokio/src/sync/batch_semaphore.rs

```diff
@@ -189,44 +188,58 @@ impl Semaphore {
     ///  
     /// If `rem` exceeds the number of permits needed by the wait list, the
     /// remainder are assigned back to the semaphore.
-    fn add_permits_locked(&self, mut rem: usize, mut waiters: MutexGuard<'_, Waitlist>) {
-        // Starting from the back of the wait queue, assign each waiter as many
-        // permits as it needs until we run out of permits to assign.
-        loop {
-            // Was the waiter assigned enough permits to wake it?
-            let woken = waiters
-                .queue
-                .last()
-                .map(|waiter| waiter.assign_permits(&mut rem))
-                .unwrap_or(false);
-            if !woken {
-                break;
+    fn add_permits_locked(&self, mut rem: usize, waiters: MutexGuard<'_, Waitlist>) {
+        let mut wakers = [None, None, None, None, None, None, None, None];
+        let mut lock = Some(waiters);
+        let mut is_empty = false;
+        while rem > 0 {
+            let mut waiters = lock.take().unwrap_or_else(|| self.waiters.lock().unwrap());
```

Comment:
Typically I'm not a big fan of releasing a lock and reacquiring it in the same method, since it often hides a concurrency bug between those times.

I don't think that's the case here, and the fact that the `MutexGuard` is passed as a parameter and can't be reused by the caller should make sure that nothing bad can happen there too. So I guess it's fine in this case :-)

Reply:
Yeah, it felt kinda ugly to me, too. But, since we can only wake a fixed number of waiters per iteration, we will need to reacquire the lock, and I thought this was better than making all the callsites repeatedly call the method.